File: PlayerModel.cpp
//      player.cpp
//

#include <cstdlib>
#include <iostream>
#include <map>
#include "sqlite3.h"
#include "PlayerModel.h"
#include "Utils.h"

using namespace std;


PlayerModel::PlayerModel()
{
}

PlayerModel* PlayerModel::loadByLoginAndPassword(string login, string password)
{
    PlayerModel *pm;
    sqlite3 *db;
    int rc;
    sqlite3_stmt *stmt = NULL;

    // Open the test.db file
    rc = sqlite3_open("database/rpg.sql", &db);

    if(rc){
        // failed
        return NULL;
    }

    string query = "SELECT id_player FROM player WHERE login = '" + login + "' AND password = '" + password + "'";

    rc = sqlite3_prepare(
        db,
        query.c_str(),
        query.size() + 1,
        &stmt,
        NULL
    );

    if (sqlite3_step(stmt) == SQLITE_ROW) {
        pm = new PlayerModel();
        pm->setPk(sqlite3_column_int(stmt, 0));
    }

    sqlite3_finalize( stmt );
    sqlite3_close( db );

    return pm;
}

bool PlayerModel::save()
{
    _playerFields["date_creation"] = "2012-03-24";

    CharacterModel::setName(_playerFields["login"]);
    CharacterModel::save();
    _playerFields["id_character"] = Utils::itos(CharacterModel::getPk());

    _setPk(Model::insert("player", _playerFields));

    return true;
}

void PlayerModel::_setPk(int pk)
{
    _playerFields["id_player"] = Utils::itos(pk);
}

void PlayerModel::setLogin(std::string login)
{
    _playerFields["login"] = login;
}

void PlayerModel::setPassword(std::string password)
{
    _playerFields["password"] = password;
}

int PlayerModel::getPk()
{
    return std::atoi(_playerFields["id_character"].c_str());
}

std::string PlayerModel::getLogin()
{
    return _playerFields["login"];
}

std::string PlayerModel::getPassword()
{
    return _playerFields["password"];
}

File: PlayerModel.py
# -*- coding: utf8 -*-

from CharacterModel import CharacterModel
from Model import Model
import datetime


class PlayerModel(CharacterModel):
    def __init__(self, idCharacter=None):
        super(PlayerModel, self).__init__(idCharacter)
        self._playerFields = dict()

    def getPk(self):
        return self._playerFields["id_player"]

    @staticmethod
    def loadByLoginAndPassword(login, password):
        query = "\
            SELECT\
                id_player,\
                login,\
                p.id_character,\
                name,\
                id_species,\
                id_gender\
            FROM\
                player AS p\
                join 'character' AS c ON p.id_character = c.id_character\
            WHERE\
                login = ? AND password = ?\
        "

        model = Model.fetchOneRow(query, (login, password))

        if len(model) > 0:
            pm = PlayerModel(model['id_character'])
            pm._setPk(model['id_player'])
            pm.setLogin(model['login'])
            pm.setIdCharacter(model['id_character'])
            return pm

        return None

    @staticmethod
    def loadByLogin(login):
        query = "\
            SELECT\
                id_player,\
                login\
            FROM\
                player\
            WHERE\
                login = ?\
        "

        model = Model.fetchOneRow(query, [login])

        if len(model) > 0:
            pm = PlayerModel(model['id_character'])
            pm._setPk(model['id_player'])
            pm.setLogin(model['login'])
            pm.setIdCharacter(model['id_character'])
            return pm

        return None

    def _setPk(self, pk):
        self._playerFields["id_player"] = pk

    def setLogin(self, login):
        self._playerFields["login"] = login

    def setIdCharacter(self, idCharacter):
        self._playerFields["id_character"] = idCharacter

    def getIdCharacter(self):
        return self._playerFields["id_character"]

    def setPassword(self, password):
        self._playerFields["password"] = password

    def save(self):
        self._playerFields["date_creation"] = datetime.datetime.now()
        self.setName(self._playerFields["login"])

        super(PlayerModel, self).save()

        self._playerFields["id_character"] = CharacterModel.getPk(self)
        self._setPk(Model.insert("player", self._playerFields))

        return True
